File: src/services/api/payment.service.ts
import { api } from '@/lib/api-client'

export interface PaymentRequest {
  amount: number
  method: string
  transactionId: string
  paymentNumber?: string
  screenshot?: string | null
  contentType?: string
  contentId?: string
  contentTitle?: string
  classLevel?: string
}

export interface PaymentItem {
  id: string
  userId: string
  amount: number
  method: string
  transactionId: string
  paymentNumber?: string | null
  screenshot?: string | null
  contentType?: string | null
  contentId?: string | null
  contentTitle?: string | null
  classLevel?: string | null
  status: string
  adminNote?: string | null
  reviewedBy?: string | null
  reviewedAt?: string | null
  createdAt: string
  updatedAt: string
  user?: { id: string; name: string | null; email: string }
}

export interface AccessCheckResult {
  purchased: boolean
  pendingPayment: boolean
  reason?: string | null
}

export interface BundleItemData {
  id: string
  contentType: string
  contentId: string
  contentTitle: string | null
  contentPrice: number
  order: number
}

export const paymentService = {
  // Create payment
  create: (data: PaymentRequest) =>
    api.post<{ message: string; payment: PaymentItem }>('payment', data),

  // List payments (authenticated user sees own, admin can see all)
  list: (params?: Record<string, string | number | boolean>) =>
    api.get<{ payments: PaymentItem[]; pagination: { page: number; limit: number; total: number; totalPages: number } }>(
      'payment',
      params,
      { retries: 1, retryDelay: 1500 }  // idempotent GET — safe to retry
    ),

  // Get payment by ID
  getById: (id: string) =>
    api.get<{ payment: PaymentItem }>(`payment/${id}`),

  // User payments
  getUserPayments: () =>
    api.get<{ data: PaymentItem[] }>(
      'user/payments',
      undefined,
      { retries: 1, retryDelay: 1500 }  // idempotent GET — safe to retry
    ),

  // Check content access (called frequently — resilient to transient failures)
  checkAccess: (contentType: string, contentId: string, userId: string) =>
    api.get<AccessCheckResult>(
      'payment/check',
      { contentType, contentId, userId },
      { retries: 1, retryDelay: 1000 }  // idempotent GET — safe to retry
    ),

  // Batch check access (subscriptions)
  batchCheckAccess: (data: { contentIds: string[]; contentType: string }) =>
    api.post<{ accessMap: Record<string, boolean> }>(
      'payment/batch-check',
      data,
      { retries: 1, retryDelay: 1500 }  // POST but safe — server checks duplicates
    ),

  // Bundle items
  getBundleItems: (bundleId: string) =>
    api.get<{ data: { items: BundleItemData[]; title: string } }>(
      `bundles/${bundleId}`,
      undefined,
      { retries: 1, retryDelay: 1500 }  // idempotent GET — safe to retry
    ),

  getBundlesForContent: (contentType: string, contentId: string) =>
    api.get<{ data: { bundles: { id: string; title: string }[] } }>('content/bundles-for', { contentType, contentId }),
}

File: download/sikkha_api_client/lib/src/services/payment_service.dart
import '../client.dart';
import '../models/models.dart';
import '../models/payment.dart';

/// Payment & Purchase API service.
class PaymentService {
  final SikkhaApiClient _client;

  PaymentService(this._client);

  /// List payments for the authenticated user.
  Future<ApiResponse<List<Payment>>> getPayments({
    int page = 1,
    int limit = 20,
    String? status,
    String? contentType,
  }) async {
    final params = <String, String?>{
      'page': page.toString(),
      'limit': limit.toString(),
      if (status != null) 'status': status,
      if (contentType != null) 'contentType': contentType,
    };

    final response = await _client.get<Map<String, dynamic>>(
      '/api/payment',
      queryParams: params,
    );
    final data = response.data as Map<String, dynamic>?;
    final payments = (data?['payments'] as List? ?? [])
        .map((e) => Payment.fromJson(e as Map<String, dynamic>))
        .toList();

    return ApiResponse<List<Payment>>(
      success: response.success,
      data: payments,
      pagination: response.pagination,
    );
  }

  /// Create a payment request.
  Future<Payment> createPayment(PaymentRequest request) async {
    final response = await _client.post<Payment>(
      '/api/payment',
      body: request.toJson(),
      requiresCsrf: true,
      fromJson: (json) => Payment.fromJson(json['payment'] as Map<String, dynamic>? ?? json),
    );
    return response.data!;
  }

  /// Get payment detail.
  Future<Payment> getPayment(String id) async {
    final response = await _client.get<Payment>(
      '/api/payment/$id',
      fromJson: Payment.fromJson,
    );
    return response.data!;
  }

  /// Check content access.
  Future<AccessCheckResult> checkAccess({
    required String contentType,
    required String contentId,
    String? userId,
  }) async {
    final params = <String, String?>{
      'contentType': contentType,
      'contentId': contentId,
      if (userId != null) 'userId': userId,
    };

    final response = await _client.get<AccessCheckResult>(
      '/api/payment/check',
      queryParams: params,
      fromJson: AccessCheckResult.fromJson,
    );
    return response.data ?? AccessCheckResult();
  }

  /// Batch check content access.
  Future<List<BatchAccessCheckItem>> batchCheckAccess(List<ContentIdentifier> items) async {
    final response = await _client.post<Map<String, dynamic>>(
      '/api/payment/batch-check',
      body: BatchAccessCheckRequest(items: items).toJson(),
      requiresCsrf: true,
    );
    final data = response.data as Map<String, dynamic>?;
    final resultItems = (data?['items'] as List? ?? [])
        .map((e) => BatchAccessCheckItem.fromJson(e as Map<String, dynamic>))
        .toList();
    return resultItems;
  }

  /// Get content info for payment dialog.
  Future<ContentInfo> getContentInfo({
    required String contentType,
    required String contentId,
    String? classLevel,
  }) async {
    final params = <String, String?>{
      'contentType': contentType,
      'contentId': contentId,
      if (classLevel != null) 'classLevel': classLevel,
    };

    final response = await _client.get<ContentInfo>(
      '/api/payment/content-info',
      queryParams: params,
      fromJson: ContentInfo.fromJson,
    );
    return response.data!;
  }

  /// Get payment account numbers.
  Future<PaymentAccounts> getAccounts() async {
    final response = await _client.get<Map<String, dynamic>>('/api/payment/accounts');
    final data = response.data as Map<String, dynamic>?;
    final accounts = data?['accounts'] as Map<String, dynamic>? ?? {};
    return PaymentAccounts.fromJson(accounts);
  }

  /// Get user's purchase history.
  Future<List<Payment>> getPurchases({int page = 1, int limit = 20}) async {
    final params = <String, String?>{
      'page': page.toString(),
      'limit': limit.toString(),
    };
    final response = await _client.get<Map<String, dynamic>>(
      '/api/payment/purchases',
      queryParams: params,
    );
    final data = response.data as Map<String, dynamic>?;
    final purchases = (data?['purchases'] as List? ?? [])
        .map((e) => Payment.fromJson(e as Map<String, dynamic>))
        .toList();
    return purchases;
  }
}
